qermit/taskgraph/mitres.py

```diff
@@ -302,3 +302,141 @@ def run(self, circuit_shots: List[CircuitShots]) -> List[BackendResult]:  # type
+        number of shots.
+        :rtype: Tuple[List[CircuitShots], List[int]]
+        """
+
+        # A new list of circuits and shots
+        split_circuit_wires = []
+        # and index of which circuits in the new list correspond to which in
+        # the original list. This is to say all circuits in the new list
+        # assigned an index i by this index list correspond to original
+        # circuit i.
+        split_index = []
+
+        # For each circuit, check if the number of shots requested is greater
+        # then the maximum, and divide the job into several smaller ones if
+        # this is the case.
+        for i, circ_shots in enumerate(circuit_wires):
+
+            shots_to_run = circ_shots.Shots
+
+            while shots_to_run > 0:
+
+                split_circ_shots = CircuitShots(
+                    circ_shots.Circuit, min(max_shots, shots_to_run)
+                )
+                # append a job with a reduced number of shots to the list.
+                split_circuit_wires.append(split_circ_shots)
+                # Append the original circuit index to the index list.
+                split_index.append(i)
+                shots_to_run -= max_shots
+
+        return (split_circuit_wires, split_index)
+
+    return MitTask(_label="SplitShots", _n_in_wires=1, _n_out_wires=2, _method=task)
+
+
+def group_shots_task_gen() -> MitTask:
+    """Returns task which groups together jobs that correspond to the same
+    circuit. Original larger job could have been split by task returned
+    by split_shots_task_gen.
+
+    :return: Task merging jobs which correspond to the same circuit.
+    :rtype: MitTask
+    """
+
+    def task(
+        obj, results_wires: List[BackendResult], split_index: List[int]
+    ) -> Tuple[List[BackendResult]]:
+        """Task which merges jobs which correspond to the same circuit.
+
+        :param results_wires: A list of results, some of which correspond to
+        the same circuit.
+        :type results_wires: List[BackendResult]
+        :param split_index: A list of indexes identifying which results
+        correspond to which circuit. Those results with index i correspond to
+        the circuit i.
+        :type split_index: List[int]
+        :return: A new list of results, where results which correspond to the
+        same circuit have been grouped together.
+        :rtype: Tuple[List[BackendResult]]
+        """
+
+        # A new list of results where some results have been merged.
+        merged_results = []
+
+        # A list of lists. Each internal list contains results that correspond
+        # to the circuit of the same index.
+        grouped_results = [
+            [result for result, i in zip(results_wires, split_index) if i == j]
+            for j in range(split_index[-1] + 1)
+        ]
+
+        # For each circuit, combine the results.
+        for fixed_circ in grouped_results:
```

Comment:
I do like this, but I'd like to suggest we go for consistency with how this is done elsewhere in the code:
```
        # For each circuit, combine the results.
        for chunk in grouped_results:
            # initialise result with the first that corresponds to this circuit.
            # circ_all_result_dict = fixed_circ[0].to_dict()
            all_counts = [result.get_counts() for result in chunk]
            combined_counts = reduce(operator.add, all_counts)

            outcome_array = {
                OutcomeArray.from_readouts([key]): val
                for key, val in combined_counts.items()
            }
            outcome_counts = Counter(outcome_array)
            merged_results.append(
                BackendResult(
                    counts=outcome_counts, c_bits=cast(Sequence[Bit], chunk[0].c_bits)
                )
            )

Reply:
yep 
only minor edit 
`all_counts = (result.get_counts() for result in chunk)`
(note parantheses)

Reply:
I tried to do the same thing with shots, just to hold onto the order of the shots. I hope that also works?